### PR TITLE
[SAP] handle DuplicateName in _migrate_attached_cross_vc

### PR DESCRIPTION
**Pull request by leust**
The diskless backing could have already been created by a
previous (failed) migration attempt. We do not clean up the
diskless backings for failed migrations, and thus we might
end up with DuplicateName error in subsequent attempts.

We attempt to reuse the existing backing by checking if it
is diskless (doesn't have any disks attached to it). Otherwise
we let cinder to update the host of the volume, but set the
migration_status to error.

Change-Id: I2e0bd27b649bd75f988a5661f83ff476020dc49b

### CODE REVIEW

File: cinder/volume/drivers/vmware/vmdk.py [hunk 1]
```diff
@@ -3179,6 +3179,14 @@ def _migrate_attached_cross_vc(self, context, dest_host, volume, backing):
                     CREATE_PARAM_DISK_LESS: True
                 })
             return (True, None)
+        except exceptions.DuplicateName:
+            # The disk_less backing could have already been created by a
```
Comment (joker-at-work) [inline cinder/volume/drivers/vmware/vmdk.py]: Why do you call it `disk_less` and not "diskless"?

**Review by joker-at-work — COMMENTED** [review]
The commit-message only tells us part of the problem but doesn't tell at all how we handle it.

How do we know its a disless backing? Would it make sense to check that fact?

File: cinder/volume/drivers/vmware/vmdk.py (hunk 1 above)
Comment (leust) [reply cinder/volume/drivers/vmware/vmdk.py]: I think was influenced by the variable `CREATE_PARAM_DISK_LESS` :)
I'll correct it with the next commit.

**Comment by leust** [comment]
> How do we know its a diskless backing

We don't know, however this operation being triggered only between different vCenters, there couldn't be a shadowVM with an attached disk at the destination (at least not one created by cinder).

I'm happy to add this check, though, if you think is necessary.

**Comment by joker-at-work** [comment]
> I'm happy to add this check, though, if you think is necessary.

I'm not sure. I feel a little unsafe just re-using whatever is there and thus would probably opt for making sure it's an empty shadow-vm. Maybe best to get consensus on this and ask the others, too.

**Comment by hemna** [comment]
I think adding the check would be ok.

**Comment by leust** [comment]
> I think adding the check would be ok.

Yes. It's already added in the latest update.

File: cinder/volume/drivers/vmware/vmdk.py [hunk 2]
```diff
@@ -3179,6 +3179,18 @@ def _migrate_attached_cross_vc(self, context, dest_host, volume, backing):
...
+            # The backing could have already been created by a
+            # previous (failed) migration attempt.
+            disk_devices = self.volumeops._get_disk_devices(backing)
+            if not disk_devices:
+                LOG.info("Reusing existing diskless backing %(volume_id)s",
+                         {'volume_id': volume['id']})
+                return (True, None)
+            else:
+                LOG.exception("Found a non-empty volume backing at the "
+                              "destination during migration.")
```
Comment (joker-at-work) [inline cinder/volume/drivers/vmware/vmdk.py]: Should we put the `volume_id` in here?

File: cinder/volume/drivers/vmware/vmdk.py [hunk 3]
```diff
@@ -3179,6 +3179,18 @@ def _migrate_attached_cross_vc(self, context, dest_host, volume, backing):
                     CREATE_PARAM_DISK_LESS: True
                 })
             return (True, None)
+        except exceptions.DuplicateName:
+            # The backing could have already been created by a
+            # previous (failed) migration attempt.
+            disk_devices = self.volumeops._get_disk_devices(backing)
```
Comment (joker-at-work) [inline cinder/volume/drivers/vmware/vmdk.py]: Why do we have a backing already that we can check, when we call `create_backing()` above to create one? I would think we would have to query the target vCenter for the already-existing shadow-VM and probably cannot use whatever `backing` points to.

Comment (leust) [reply cinder/volume/drivers/vmware/vmdk.py]: Yes, my bad, I have mislooked at it. Will rework it.

Comment (leust) [reply cinder/volume/drivers/vmware/vmdk.py]: I moved the handling to `_create_backing` where the actual creation happens (at the destination).

**Review by joker-at-work — APPROVED** [review]
What happens if multiple requests to create a diskless backing are queued up? The first one would go through, creating the diskless backing, returning it. The second one will find the (still diskless) backing created by the first one.

Who calls the `_create_backing()` with `disk_less`? Looks like it's `_create_volume_from_non_stream_optimized_image()` and `_manage_existing_int()` (which is e.g. called from `_clone_attached_volume()`). The first one is probably fine, as it should™ fill the backing immediately before anybody else can return. The second one should only come up with a new volume id and also directly adds a disk to the backing, so that's probably fine, too.

I think this is fine, even with the context expanded to `_create_backing()` instead of only `_migrate_attached_cross_vc()`.

**Pull request merged**